File: scripts/gen-scene-04-roadmap.py
#!/usr/bin/env python3
"""Generate scene-04-roadmap using Vertex AI Imagen 3 with character references."""

import warnings
warnings.filterwarnings("ignore")

import sys
import os
import base64
import io
import json
import subprocess

CHAR_DIR = "/Users/guru/Desktop/Projects/Reconnect/reconnect_website/public/images/train-journey/characters"
OUT_PATH = "/Users/guru/Desktop/Projects/Reconnect/reconnect_website/public/images/train-journey/scene-04-roadmap.png"
PROJECT = "radlabs-497004"
LOCATION = "us-central1"
ENDPOINT = f"https://{LOCATION}-aiplatform.googleapis.com/v1/projects/{PROJECT}/locations/{LOCATION}/publishers/google/models/imagen-3.0-capability-001:predict"
FALLBACK_ENDPOINT = f"https://{LOCATION}-aiplatform.googleapis.com/v1/projects/{PROJECT}/locations/{LOCATION}/publishers/google/models/imagen-3.0-generate-001:predict"

PROMPT = """Editorial illustration in the spirit of Christoph Niemann meets Malika Favre, with the restraint of Tom Gauld. New Yorker cover aesthetic, Bloomberg Businessweek, Apple editorial.

PALETTE (strict):
- Background: warm cream #FAF7F0
- Primary fills: deep medical blue #2563EB
- Ink/lines: charcoal #1A1A1A
- Accent: muted coral #E76F51 (5% max, only on roadmap calendar icon)

TECHNIQUE: Flat 2D only. No gradients, no shadows, no rendering. Confident varied linework 1-4pt. Subtle riso grain texture on fills. Generous negative space. Adult proportions, dignified.

SCENE: Interior cross-section of train conference car (car #4). Five people seated around a long rectangular conference table centered in frame.

LEFT SIDE of table: Dr. Shruthi — Indian woman, white coat, short dark hair, glasses, warm confident expression, facing slightly right. Next to her: Nutritionist — South Indian woman, bright professional attire, gentle smile.

RIGHT SIDE of table: Trainer — athletic Indian man, sports/casual attire, engaged posture. Next to him: Physiotherapist — Indian woman, professional calm expression.

HEAD OF TABLE (center-right, facing viewer): Patient — Indian woman mid-40s, elegant kurta/attire, mid-laugh with head tilted slightly back, eyes crinkled, genuine joy. The whole team is smiling with her in collegial warmth.

ON THE TABLE: A large unfolded paper roadmap. On it, three icons connected by a dashed dotted line:
(1) Salad bowl icon — simple flat illustration
(2) Wristwatch icon with "10K" text on face
(3) Calendar grid icon with one square circled in coral

TRAIN CAR INTERIOR: Four rectangular windows along the top wall showing simple sky silhouette outside. Overhead strip lighting as flat rectangles. Cream wall panels. Blue seat cushions.

TRACK: Visible beneath the train, curved gently upward to the right.

LABEL: At top of composition, clean sans-serif text: "04 — STRUCTURED ROADMAP" in charcoal.

16:9 landscape. Cream background. Beautiful flat editorial illustration."""


def compress_image(path: str, max_size: int = 512) -> str:
    """Compress image to JPEG and return base64."""
    try:
        from PIL import Image
        img = Image.open(path).convert("RGB")
        img.thumbnail((max_size, max_size))
        buf = io.BytesIO()
        img.save(buf, format="JPEG", quality=85)
        return base64.b64encode(buf.getvalue()).decode()
    except ImportError:
        # Fallback: use raw PNG bytes
        with open(path, "rb") as f:
            return base64.b64encode(f.read()).decode()


def get_token() -> str:
    return subprocess.check_output(
        ["gcloud", "auth", "print-access-token"], text=True
    ).strip()


def call_api(endpoint: str, body: dict, token: str):
    import urllib.request
    data = json.dumps(body).encode("utf-8")
    req = urllib.request.Request(
        endpoint,
        data=data,
        headers={
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=180) as resp:
        return json.loads(resp.read().decode("utf-8"))


def main():
    print(f"Getting auth token...")
    token = get_token()

    # Load and compress character reference images
    # NOTE: Imagen 3 capability-001 supports max 4 refs for square (1:1) and max 2 for 16:9.
    # For 16:9, use only the 2 most important characters.
    char_files = [
        ("char-dr-shruthi.png", "Dr. Shruthi: Indian female doctor, white coat, short dark hair, glasses"),
        ("char-patient.png", "Patient: Indian woman mid-40s, warm smile, elegant attire"),
    ]

    print(f"Loading and compressing {len(char_files)} character references...")
    reference_images = []
    for i, (fname, description) in enumerate(char_files):
        path = os.path.join(CHAR_DIR, fname)
        if not os.path.exists(path):
            print(f"  WARNING: {path} not found, skipping")
            continue
        b64 = compress_image(path, max_size=512)
        reference_images.append({
            "referenceId": i + 1,
            "referenceType": "REFERENCE_TYPE_SUBJECT",
            "referenceImage": {
                "bytesBase64Encoded": b64,
            },
            "subjectImageConfig": {
                "subjectType": "SUBJECT_TYPE_PERSON",
                "subjectDescription": description,
            },
        })
        print(f"  Loaded {fname}: {len(b64):,} chars")

    print(f"\nGenerating with {len(reference_images)} character references...")
    print("Using model: imagen-3.0-capability-001 (Imagen 3 with editing/reference support)")

    body = {
        "instances": [{
            "prompt": PROMPT,
            "referenceImages": reference_images,
        }],
        "parameters": {
            "sampleCount": 1,
            "aspectRatio": "16:9",
            "personGeneration": "allow_adult",
            "safetyFilterLevel": "block_some",
            "addWatermark": False,
        },
    }

    try:
        result = call_api(ENDPOINT, body, token)
        predictions = result.get("predictions", [])
        if predictions and predictions[0].get("bytesBase64Encoded"):
            b64_out = predictions[0]["bytesBase64Encoded"]
            with open(OUT_PATH, "wb") as f:
                f.write(base64.b64decode(b64_out))
            size = os.path.getsize(OUT_PATH)
            print(f"\nSaved: {OUT_PATH}")
            print(f"File size: {size:,} bytes")
            print("Method: vertex-ai with character references (imagen-3.0-capability-001)")
            return
        else:
            print(f"No image in response: {json.dumps(result)[:300]}")
            raise ValueError("No image in response")

    except Exception as e:
        print(f"\nError with reference model: {e}")
        print("Falling back to imagen-3.0-generate-001 (text-only, best quality prompt)...")

        fallback_body = {
            "instances": [{"prompt": PROMPT}],
            "parameters": {
                "sampleCount": 1,
                "aspectRatio": "16:9",
                "personGeneration": "allow_adult",
                "safetyFilterLevel": "block_some",
                "addWatermark": False,
            },
        }

        try:
            result = call_api(FALLBACK_ENDPOINT, fallback_body, token)
            predictions = result.get("predictions", [])
            if predictions and predictions[0].get("bytesBase64Encoded"):
                b64_out = predictions[0]["bytesBase64Encoded"]
                with open(OUT_PATH, "wb") as f:
                    f.write(base64.b64decode(b64_out))
                size = os.path.getsize(OUT_PATH)
                print(f"\nSaved (text-only fallback): {OUT_PATH}")
                print(f"File size: {size:,} bytes")
                return
        except Exception as e2:
            print(f"Fallback also failed: {e2}")
            sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/gen-scene-04-roadmap.mjs
import { execSync } from "child_process";
import { writeFileSync, readFileSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __dir = dirname(fileURLToPath(import.meta.url));
const CHAR_DIR = join(__dir, "../public/images/train-journey/characters");
const OUT_PATH = join(__dir, "../public/images/train-journey/scene-04-roadmap.png");

const PROJECT = "radlabs-497004";
const LOCATION = "us-central1";

// Imagen 3 generate endpoint
const ENDPOINT = `https://${LOCATION}-aiplatform.googleapis.com/v1/projects/${PROJECT}/locations/${LOCATION}/publishers/google/models/imagen-3.0-generate-001:predict`;

const PROMPT = `Editorial illustration in the spirit of Christoph Niemann meets Malika Favre, with the restraint of Tom Gauld. Reminiscent of New Yorker covers, Bloomberg Businessweek, Apple editorial work.

PALETTE (strict):
- Primary: deep medical blue #2563EB
- Background: warm cream #FAF7F0
- Accent: muted coral #E76F51 (max 5% of canvas)
- Ink: charcoal #1A1A1A

TECHNIQUE:
- Flat 2D, no gradients, no rendering, no shading
- Confident line work with intentional weight variation 1pt to 4pt
- Textured fills with subtle riso-print grain
- Generous negative space, formal composition
- All figures drawn with adult proportions and dignity

TONE: intelligent, quietly emotional, slightly witty, never sentimental.

AVOID: cartoonish, chibi, anime, children's book style, big eyes, sweat drops, exclamation marks, speech bubbles, 3D rendering, gradients, Corporate Memphis, generic stock illustration, photorealism.

SCENE: Train car four interior shown in cross-section. A conference car. Four healthcare professionals and a patient seated around a long rectangular table. Left to right: Dr. Shruthi (Indian female doctor in white coat, short dark hair, confident posture), a nutritionist (female, warm expression), a trainer (athletic build, male), a physiotherapist (female, professional attire). The patient (Indian woman, mid-40s, saree or elegant kurta) sits at the head of the table, genuinely smiling mid-laugh, head tilted slightly back. The entire team smiles warmly with her — collegial, genuine energy, not clinical.

On the table surface visible to viewer: a folded paper roadmap, unfolded to show three illustrated icons connected by a dotted line path:
1. A salad bowl (nutrition icon)
2. A wristwatch with "10K" subtly on face (movement icon)
3. A small calendar grid with one square circled in coral (consistency icon)

Train car interior: rectangular windows showing passing landscape, overhead lighting as simple rectangular shapes, warm cream walls. Track underneath curves gently upward toward the right.

Label at top: "04 — STRUCTURED ROADMAP" in clean sans-serif, charcoal ink.

16:9 aspect ratio. Warm cream #FAF7F0 background. Flat editorial illustration. Beautiful, precise, intentional.`;

async function main() {
  console.log("Getting auth token...");
  const token = execSync("gcloud auth print-access-token", { encoding: "utf8" }).trim();

  // Read character reference images as base64
  console.log("Loading character reference images...");
  const chars = [
    { name: "char-dr-shruthi", file: "char-dr-shruthi.png" },
    { name: "char-patient", file: "char-patient.png" },
    { name: "char-nutritionist", file: "char-nutritionist.png" },
    { name: "char-trainer", file: "char-trainer.png" },
    { name: "char-physio", file: "char-physio.png" },
  ];

  // Imagen 3 allows max 4 reference images — prioritise the two key characters
  // plus nutritionist and physio (trainer described well enough via text)
  const priorityChars = chars.filter((c) => c.name !== "char-trainer");

  // Build subject reference images for Imagen 3.
  // Correct REST field: "referenceImage" (NOT "image") containing bytesBase64Encoded.
  // Images must be compressed to JPEG to keep payload manageable.
  const subjectReferenceImages = await Promise.all(
    priorityChars.map(async (c, i) => {
      // Use sharp if available, otherwise load raw bytes
      const imgData = readFileSync(join(CHAR_DIR, c.file));
      // Compress via jimp or just use raw bytes (resize not needed since JPEG is fine)
      return {
        referenceId: i + 1,
        referenceType: "REFERENCE_TYPE_SUBJECT",
        referenceImage: {
          bytesBase64Encoded: imgData.toString("base64"),
        },
        subjectImageConfig: {
          subjectType: "SUBJECT_TYPE_PERSON",
          subjectDescription: "",
        },
      };
    })
  );

  console.log(`Loaded ${subjectReferenceImages.length} character references (max 4 allowed by API)`);

  const requestBody = {
    instances: [
      {
        prompt: PROMPT,
        referenceImages: subjectReferenceImages,
      },
    ],
    parameters: {
      sampleCount: 1,
      aspectRatio: "16:9",
      safetyFilterLevel: "block_some",
      personGeneration: "allow_adult",
      addWatermark: false,
    },
  };

  console.log("Calling Vertex AI Imagen 3 with character references...");
  const resp = await fetch(ENDPOINT, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(requestBody),
  });

  if (!resp.ok) {
    const errText = await resp.text();
    console.error(`HTTP ${resp.status}: ${errText.slice(0, 500)}`);

    // Fallback: try without reference images (plain text-to-image)
    console.log("\nFalling back to text-only generation...");
    await generateTextOnly(token);
    return;
  }

  const json = await resp.json();
  const b64 = json?.predictions?.[0]?.bytesBase64Encoded;
  if (!b64) {
    console.error("No image in response:", JSON.stringify(json).slice(0, 300));
    console.log("\nFalling back to text-only generation...");
    await generateTextOnly(token);
    return;
  }

  writeFileSync(OUT_PATH, Buffer.from(b64, "base64"));
  console.log(`\nSaved: ${OUT_PATH}`);
}

async function generateTextOnly(token) {
  const fallbackPrompt = `${PROMPT}

Characters to include:
- Dr. Shruthi: Indian female doctor, white coat, short dark hair, glasses, confident professional bearing
- Patient: Indian woman mid-40s, warm smile, head tilted back in genuine laughter, elegant attire
- Nutritionist: South Indian woman, warm expression, professional casual
- Trainer: Athletic Indian man, energetic posture, sports-adjacent attire
- Physiotherapist: Indian woman, calm focused expression, medical professional look

All five characters around conference table inside train car.`;

  const resp = await fetch(ENDPOINT, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      instances: [{ prompt: fallbackPrompt }],
      parameters: {
        sampleCount: 1,
        aspectRatio: "16:9",
        safetyFilterLevel: "block_some",
        personGeneration: "allow_adult",
        addWatermark: false,
      },
    }),
  });

  if (!resp.ok) {
    const errText = await resp.text();
    throw new Error(`HTTP ${resp.status}: ${errText.slice(0, 500)}`);
  }

  const json = await resp.json();
  const b64 = json?.predictions?.[0]?.bytesBase64Encoded;
  if (!b64) throw new Error("No image in response: " + JSON.stringify(json).slice(0, 300));

  writeFileSync(OUT_PATH, Buffer.from(b64, "base64"));
  console.log(`Saved (text-only fallback): ${OUT_PATH}`);
}

main().catch((e) => {
  console.error("Fatal error:", e.message);
  process.exit(1);
});
